### PR TITLE
YamlCreate Improvements

### LINKED ISSUE #178: [New Feature]: Decode URLs
### Description of the new feature/enhancement

Lots of urls are encoded and should be decoded in YamlCreate:

Encoded:
`https://github.com/standardnotes/app/releases/download/%40standardnotes%2Fdesktop%403.158.4/standard-notes-3.158.4-win-x64.exe`

Decoded:
`https://github.com/standardnotes/app/releases/download/@standardnotes/desktop@3.158.4/standard-notes-3.158.4-win-x64.exe`

Note `%40` becoming `@` and `%2F` becoming `/`.

### Proposed technical implementation details (optional)

```pwsh
Add-Type -AssemblyName System.Web
$url = "https%3A%2F%2Fwww.example.com%2F"
$decodedUrl = [System.Web.HttpUtility]::UrlDecode($url)
Write-Output $decodedUrl
```

### LINKED ISSUE #181: [New Feature]: Identify scope from URL
### Description of the new feature/enhancement

YamlCreate should check for the key words "user" or "machine" in a URL and skip the scope prompt accordingly.

Komac's implementation:
```kotlin
fun Url.findScope(): InstallerManifest.Scope? {
    return when {
        fullPath.contains(other = "user", ignoreCase = true) -> InstallerManifest.Scope.User
        fullPath.contains(other = "machine", ignoreCase = true) -> InstallerManifest.Scope.Machine
        else -> null
    }
}
```

### PR DESCRIPTION
**Pull request by Trenly**
- [x] Have you signed the [Contributor License Agreement](https://cla.opensource.microsoft.com/microsoft/winget-pkgs)?

* Resolves Trenly/winget-pkgs#178
* Resolves Trenly/winget-pkgs#181

cc @denelon @russellbanks @mdanish-kh

 ###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/microsoft/winget-pkgs/pull/106351)

### CODE REVIEW

**Comment by russellbanks** [comment]
Depending on the implementation, `+` can sometimes get encoded as a space. Quickly testing this PowerShell code shows this happening to this too:
```pwsh
Add-Type -AssemblyName System.Web
$url = "https://github.com/adoptium/temurin17-binaries/releases/download/jdk-17.0.7+7/OpenJDK17U-jdk_x64_windows_hotspot_17.0.7_7.msi"
$decodedUrl = [System.Web.HttpUtility]::UrlDecode($url)
Write-Output $decodedUrl
```
```
https://github.com/adoptium/temurin17-binaries/releases/download/jdk-17.0.7 7/OpenJDK17U-jdk_x64_windows_hotspot_17.0.7_7.msi
```

**Comment by Trenly** [comment]
> Depending on the implementation, `+` can sometimes get decoded as a space. Quickly testing this PowerShell code shows this happening to this too:
> 
> ```powershell
> Add-Type -AssemblyName System.Web
> $url = "https://github.com/adoptium/temurin17-binaries/releases/download/jdk-17.0.7+7/OpenJDK17U-jdk_x64_windows_hotspot_17.0.7_7.msi"
> $decodedUrl = [System.Web.HttpUtility]::UrlDecode($url)
> Write-Output $decodedUrl
> ```
> 
> ```
> https://github.com/adoptium/temurin17-binaries/releases/download/jdk-17.0.7 7/OpenJDK17U-jdk_x64_windows_hotspot_17.0.7_7.msi
> ```

:o 

Any other suggestions on what code to use then, other than the built in decode?

**Comment by denelon** [comment]
@SteveL-MSFT,
Any suggestions?

**Comment by Trenly** [comment]
> The easiest workaround without a different implementation would be to replace `+` with `%2F` before decoding, which will then get decoded back into a `+`.
> 
> This is the only instance of this behaviour because this is intentional when working with URL encoded forms.
> 
> ```powershell
> Add-Type -AssemblyName System.Web
> $url = "https://github.com/adoptium/temurin17-binaries/releases/download/jdk-17.0.7+7/OpenJDK17U-jdk_x64_windows_hotspot_17.0.7_7.msi"
> $decodedUrl = [System.Web.HttpUtility]::UrlDecode($url.Replace('+','%2B'))
> Write-Output $decodedUrl
> ```

Yep, just encode the URL before Decoding

**Comment by russellbanks** [comment]
> Yep, just encode the URL before Decoding

Encoding the URL is unnecessary when you can just replace `+` with `%2B` as this doesn't happen with any other character

File: Tools/YamlCreate.ps1 [hunk 1]
```diff
@@ -2686,6 +2703,7 @@ Switch ($script:Option) {
     Write-Host 'Updating Manifest Information. This may take a while...' -ForegroundColor Blue
     $_NewInstallers = @();
     foreach ($_Installer in $script:OldInstallerManifest.Installers) {
+      $_Installer['InstallerUrl'] = [System.Web.HttpUtility]::UrlDecode([System.Web.HttpUtility]::UrlEncode($_Installer.InstallerUrl))
```
Comment (russellbanks) [inline Tools/YamlCreate.ps1:2706]:
```suggestion
      $_Installer['InstallerUrl'] = [System.Web.HttpUtility]::UrlDecode($_Installer.InstallerUrl.Replace('+','%2B'))
```

File: Tools/YamlCreate.ps1 [hunk 2]
```diff
@@ -453,6 +453,7 @@ Function Request-InstallerUrl {
           }
         }
       }
+      $NewInstallerUrl = [System.Web.HttpUtility]::UrlDecode([System.Web.HttpUtility]::UrlEncode($NewInstallerUrl))
```
Comment (russellbanks) [inline Tools/YamlCreate.ps1:456]:
```suggestion
      $NewInstallerUrl = [System.Web.HttpUtility]::UrlDecode($NewInstallerUrl.Replace('+','%2B'))
```

**Review by russellbanks — APPROVED** [review]

**Review by denelon — APPROVED** [review]

**Pull request merged**